server/plugin/subscribe.go

```diff
@@ -0,0 +1,307 @@
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+
+	"github.com/nathanaelhoun/mattermost-plugin-circleci/server/store"
+)
+
+const (
+	subscribeTrigger = "subscription"
+	subscribeHint    = "<" + subscribeListTrigger + "|" +
+		subscribeChannelTrigger + "|" +
+		subscribeUnsubscribeChannelTrigger + "|" +
+		subscribeListAllChannelsTrigger + ">"
+	subscribeHelpText = "Manage your subscriptions"
+
+	subscribeListTrigger  = "list"
+	subscribeListHint     = ""
+	subscribeListHelpText = "List the CircleCI subscriptions for the current channel"
+
+	subscribeChannelTrigger  = "add"
+	subscribeChannelHint     = "[--flags]"
+	subscribeChannelHelpText = "Subscribe the current channel to CircleCI notifications for a project"
+
+	subscribeUnsubscribeChannelTrigger  = "remove"
+	subscribeUnsubscribeChannelHint     = "[--flags]"
+	subscribeUnsubscribeChannelHelpText = "Unsubscribe the current channel to CircleCI notifications for a project"
+
+	subscribeListAllChannelsTrigger  = "list-channels"
+	subscribeListAllChannelsHint     = ""
+	subscribeListAllChannelsHelpText = "List all channels in the current team subscribed to a project"
+)
+
+func getSubscribeAutoCompleteData() *model.AutocompleteData {
+	subscribe := model.NewAutocompleteData(subscribeTrigger, subscribeHint, subscribeHelpText)
+
+	subscribeList := model.NewAutocompleteData(subscribeListTrigger, subscribeListHint, subscribeListHelpText)
+
+	subscribeChannel := model.NewAutocompleteData(subscribeChannelTrigger, subscribeChannelHint, subscribeChannelHelpText)
+	subscribeChannel.AddNamedTextArgument(store.FlagOnlyFailedJobs, "Only receive notifications for failed builds", "[write anything to continue autocomplete]", "", false) // TODO use boolean flag when then are available. See https://github.com/mattermost/mattermost-server/pull/14810
+	subscribeChannel.AddNamedDynamicListArgument(namedArgProjectName, namedArgProjectHelpText, routeAutocomplete+subrouteFollowedProjects, false)
+
+	unsubscribeChannel := model.NewAutocompleteData(subscribeUnsubscribeChannelTrigger, subscribeUnsubscribeChannelHint, subscribeUnsubscribeChannelHelpText)
+	unsubscribeChannel.AddNamedDynamicListArgument(namedArgProjectName, namedArgProjectHelpText, routeAutocomplete+subrouteFollowedProjects, false)
+
+	listAllSubscribedChannels := model.NewAutocompleteData(subscribeListAllChannelsTrigger, subscribeListAllChannelsHint, subscribeListAllChannelsHelpText)
+	listAllSubscribedChannels.AddNamedDynamicListArgument(namedArgProjectName, namedArgProjectHelpText, routeAutocomplete+subrouteFollowedProjects, false)
+
+	subscribe.AddCommand(subscribeChannel)
+	subscribe.AddCommand(unsubscribeChannel)
+	subscribe.AddCommand(subscribeList)
+	subscribe.AddCommand(listAllSubscribedChannels)
+
+	return subscribe
+}
+
+func (p *Plugin) executeSubscribe(context *model.CommandArgs, circleciToken string, project *store.ProjectIdentifier, split []string) (*model.CommandResponse, *model.AppError) {
```

Comment:
```suggestion
func (p *Plugin) executeSubscribe(args *model.CommandArgs, circleciToken string, project *store.ProjectIdentifier, split []string) (*model.CommandResponse, *model.AppError) {
```
for consistency, I'd rename this `args` to match all the other execute... functions.

I was wondering if it would be useful to create a type that could be passed around to the functions.  It seems that args, circleciToken, project, and sometimes split are shared as inputs to many functions.

Reply:
If I grep the execute functions, `rg 'func \(p \*Plugin\) execute'` I see some discrepancy in variable names.  using an object type would make everything consistent across all the functions and reduce #'s of inputs.

Reply:
I have renamed this :-)
Making an object is a good idea, I keep this conversation open to remind of this!